opm/grid/cpgrid/CpGrid.cpp

```diff
@@ -1516,610 +2031,1764 @@ void CpGrid::addLgrsUpdateLeafView(const std::vector<std::array<int,3>>& cells_p
+                                                   std::unordered_map<int,std::array<int,2>> adaptedCell_to_elemLgrAndElemLgrCell,
+                                                   const int cell_count)
+{
+    const int& startingGridIdx = static_cast<int>(this->data_.size()) -1;
+    adapted_child_to_parent_cells.resize(cell_count, std::array<int,2>{-1,-1}); //  Entry is {-1,-1} when cell has no father, otherwise, {level parent cell, parent cell index}.
+    adapted_cell_to_idxInParentCell.resize(cell_count, -1);
+    // Rewrite only the entries of adapted cells that have a parent cell
+    for (int cell = 0; cell < cell_count; ++cell) {
+        const auto& elemLgr = adaptedCell_to_elemLgrAndElemLgrCell[cell][0];
+        // (elemLgr == -1) represents that the adapted cell is equivalent to a cell from the starting grid (current_view_data_)
+        if (elemLgr != -1) {
+            // Search for the level where the parent cell was born, and its index in that level grid.
+            // Notice that elemLgr is a cell index in current_view_data_ (the starting grid where elements got marked for (further) refinement).
+            const auto& element = Dune::cpgrid::Entity<0>(*current_view_data_, elemLgr, true); // elemLgr == parent cell index in starting grid.
+            const auto& elemLevel = element.level(); // parent cell level (where the parent cell was born, might not coincide with the starting grid).
+            const auto& elemLevelIdx = (elemLevel == 0) ? element.getOrigin().index() : element.getLevelElem().index(); // parent cell index in the level it was born.
+            adapted_child_to_parent_cells[cell] = {elemLevel, elemLevelIdx};
+            adapted_cell_to_idxInParentCell[cell] = adaptedCell_to_elemLgrAndElemLgrCell[cell][1];
+        }
+        else {
+            // Populate child_to_parent_ relationships from the previous (existing grids before calling adapt).
+            assert(elemLgr == -1);
+            const auto& elemIdx = adaptedCell_to_elemLgrAndElemLgrCell[cell][1];
+            // Search for the level where the parent cell was born, and its index in that level grid.
+            // Notice that elemLgr is a cell index in current_view_data_ (the starting grid where elements got marked for (further) refinement).
+            const auto& element = Dune::cpgrid::Entity<0>(*current_view_data_, elemIdx, true); // elemIdx of equivalent cell in the starting grid.
+            if(element.hasFather()) {
+                adapted_child_to_parent_cells[cell] = {element.father().level(), element.father().index()};
+                const auto& elemLevel = element.level(); // parent cell level (level where the parent cell was born, might not coincide with the starting grid).
+                const auto& elemLevelIdx = (elemLevel == 0) ? element.getOrigin().index() : element.getLevelElem().index(); // parent cell index in the level it was born.
```

Comment:
See above for a brief description of what getOrigin() and getelvelElem() do. Now, there is Entity::getEquivLevelElem() which does not throw, returns the equivalent entity in the level it was born, for all levels, including 0. 

(Maybe it's worth it to say that getLevelElem() assumes that the grid is the leaf-grid-view (it throws otherwise, that means it throws when the grid associated with the current entity is any level grid, including level 0).